crates/ironrdp-pdu/README.md

```diff
@@ -130,3 +196,129 @@ When hiding some fields is really required, one of the following approach is sug
 [2]: https://doc.rust-lang.org/reference/expressions/struct-expr.html#functional-update-syntax
 [3]: https://matklad.github.io/2022/05/29/builder-lite.html
 [4]: https://xaeroxe.github.io/init-struct-pattern/
+
+## Enumeration-like types should allow resilient parsing
+
+Network protocols are generally designed with forward and backward compatibility in mind. Items
+like status codes, error codes, and other enumeration-like types may later get extended to include
+new values. Additionally, it's not uncommon for different implementations of the same protocol to
+disagree with each other. Some implementations may exhibit a bug, inserting an unexpected value,
+while others may intentionally extend the original protocol with a custom value.
+
+Therefore, implementations of such protocols should decode these into a more flexible data type
+that can accommodate a broader range of values, even when working with a language like Rust. As
+long as the unknown value is not critical and can be handled gracefully, the implementation should
+make every effort not to fail, especially during parsing.
+
+For example, let’s consider the following `enum` in Rust:
+
+```rust
+#[repr(u32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MyNetworkCode {
+    FirstValue = 0,
+    SecondValue = 1,
+}
+```
+
+This type cannot be used to hold new values that may be added to the protocol in the future. Thus,
+it is not a suitable option.
+
+Many Rust developers will instinctively write the following instead:
+
+```rust
+#[repr(u32)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MyNetworkCode {
+    FirstValue = 0,
+    SecondValue = 1,
+    // Fallback variant
+    Unknown(u32),
+}
+```
+
+This type can indeed hold additional values in its fallback variant, and the implementation can
+remain compatible with future protocol versions as desired.
+
+However, this solution is not without its issues. There is a hard to catch forward-compatibility
+hazard at the library level: when a new value, such as `ThirdValue = 2`, is added to the enum, the
+`Unknown` variant for this value will no longer be emitted by the library. This can lead to _silent_
```

Comment:
I don't understand what it means for a value to be "emitted" by the library.

Reply:
I meant something like "the library does not construct and return the value".

To illustrate:

```rust
// library code

#[repr(u32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MyNetworkCode {
    FirstValue = 0,
    SecondValue = 1,
    Unknown(u32),
}

impl MyNetworkCode {
    fn parse_network_code(value: u32) -> MyNetworkCode {
        match value {
            0 => MyNetworkCode::FirstValue,
            1 => MyNetworkCode::SecondValue,
            _ => MyNetworkCode::Unknown(value),
        }
    }
}

// user code

fn handle_network_code(reader: /* … */) {
    let value = reader.read_u32();
    let code = MyNetworkCode::from_u32(value);
    
    if code == MyNetworkCode::Unknown(2) {
        // The library doesn’t know about this value yet, but we need to handle it because X
    }
}
```

Once the library is updated to handle the third value we get:

```rust
// library code

#[repr(u32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MyNetworkCode {
    FirstValue = 0,
    SecondValue = 1,
    ThirdValue = 2, // NEW
    Unknown(u32),
}

impl MyNetworkCode {
    fn parse_network_code(value: u32) -> MyNetworkCode {
        match value {
            0 => MyNetworkCode::FirstValue,
            1 => MyNetworkCode::SecondValue,
            2 => MyNetworkCode::ThirdValue, // NEW
            _ => MyNetworkCode::Unknown(value),
        }
    }
}

// user code (unchanged)

fn handle_network_code(reader: /* … */) {
    let value = reader.read_u32();
    let code = MyNetworkCode::from_u32(value);
    
    if code == MyNetworkCode::Unknown(2) { // <- the library does not construct this value as it used to…
        // …  the special case is not handled anymore; no warning and no error
        // was emitted by the compiler, so it’s very easy to overlook this in practice
    }
}
```

The bottom line is to avoid types which may represent the same thing in two different ways

Reply:
I see, thanks for the example. Maybe it's worth adding a TL;DR to the top of that section (rather than at the end), otherwise all looks good to me.

Reply:
You are right!